kep/61-real-load-aware-scheduling/README.md

```diff
@@ -0,0 +1,608 @@
+
+
+### BestFitBinPack Plugin
+
+
+#### Score plugin
+
+This plugin would extend the Score extension point. K8s scheduler framework calls the Score function for each node separately when scheduling a pod.
+
+Following is the algorithm:
+
+**Algorithm**
+
+1. Get the utilization of the current node to be scored. Call it A.
+2. Calculate the current pod's total CPU requests and overhead. Call it B.
+3. Calculate the expected utilization if the pod is scheduled under this node by adding i.e. U = A + B.
+4. If U &lt;= 50%, return U+50 as the score
+5. If 50% &lt; U &lt;= 100%, return 100 - U
+6. If U > 100%, return 0
+
+For example, let’s say we have three nodes X, Y, and Z, with four cores each and utilization 1, 2, and 3 cores respectively. For simplicity, let’s assume our pod to be scheduled has 0 cores CPU requests and overhead.
+
+Utilization of each node:
+
+
+```
+Ux → (1/4)*100 = 25
+Uy → (2/4)*100 = 50
+Uz → (3/4)*100 = 75
+```
+
+
+The score of each node :
+
+
+```
+X → 50 + 25 =  75
+Y → 50 + 50 = 100
+Z → 75 - 50 = 25
+```
+
+
+In the algorithm above, 50% is the target utilization rate we want to achieve on all nodes. We can be less aggressive by reducing it to 40% so that it has much lesser chances of going over 50% during spikes or unexpected loads. 
+
+In the 2nd step of the algorithm, one variant uses the current pod's total CPU limits instead of requests, to have a stronger upper bound of expected utilization.
+
+The 50% threshold value for utilisation will be made configurable via plugin argument.
+
+**Algorithm Analysis**
+
+<img src="images/bfbp-graph.png" alt="bfbp-graph" width="600" height="400"/>
+
+
+The above is a plot of the piecewise function outlined in the algorithm. The key observations here are manifold:
+
+
+
+1. As the utilization goes from 0 to 50%, we pack pods on those nodes by favoring them.
+2. The nodes are penalized linearly as the utilization goes beyond 50%, to spread the pod amongst those "hot" nodes.
+3. The positive slope begins from 50 and not from 0 because the range of score is from 0-100, and we want to maximize our score output for nodes we wish to favor so that the score is more substantial and other plugins do not affect it much.
+4. There is a break in the graph with a high drop due to the "penalty" we have.
+
+#### Spreading with BestFitBinPack
+
+
+To meet the 2nd constraint in the problem statement, we utilize the existing PodTopologySpread (PTS) plugin provided by K8s. 
+This is to avoid baking in another spreading algorithm within BFBP when there is an existing in-tree plugin and conflict our 4th goal.  
+It was proved with experiments that using a dynamic weight-based approach (changing weights of PTS and BFBP in every scheduling cycle) is not a good idea, neither it is supported by Kubernetes. 
+However, if we choose a static large weight for PTS, say 100, we can meet our goals. 
```

Comment:
why 100 instead of the default (2)? Have you experimented with 1.19?

Reply:
We _did_ experiment the final scoring function with 1.19, not the above part of assigning a static weight. It may be 2 for all we know, it doesn't matter as it is just an example.

Reply:
It does matter, to justify merging the scores in a single plugin

Reply:
Not sure if I follow you, for justification part, please follow this discussion https://github.com/kubernetes-sigs/scheduler-plugins/pull/61#discussion_r505098463 
Since we are not proposing to use a static weight in this, Even if I do the experiment again for static weight with 1.19, I will just update "100" value above to what works - say "2". It doesn't change the proposal idea.

Reply:
@alculquicondor Do you believe the value of "x" could be much lower in 1.19?

Reply:
yes

Reply:
Okay, I have removed the "100" value, and specified the extended algorithm would be configurable to turn off and use static weights. Since experiments will take a while, I would add what value we recommend if using static configuration in a later PR and want this PR merged. @Huang-Wei @alculquicondor

Reply:
The problem is not so much about the value that you suggest, but the fact that you are merging 2 plugins into one. I don't think there is justification for that, as we have already pointed out in https://github.com/kubernetes-sigs/scheduler-plugins/pull/61#discussion_r505098463